server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java

```diff
@@ -0,0 +1,128 @@
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault;
+
+import java.io.InputStream;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailrepository.api.MailKey;
+import org.apache.james.mailrepository.api.MailRepository;
+import org.apache.james.mailrepository.api.MailRepositoryStore;
+import org.apache.james.mailrepository.api.MailRepositoryUrl;
+import org.apache.james.server.core.MimeMessageInputStream;
+import org.apache.james.util.streams.Iterators;
+import org.apache.james.vault.search.Query;
+import org.reactivestreams.Publisher;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class MailRepositoryDeletedMessageVault implements DeletedMessageVault {
+    public static class Configuration {
+        private final MailRepositoryUrl urlPrefix;
+
+        public Configuration(MailRepositoryUrl urlPrefix) {
+            this.urlPrefix = urlPrefix;
+        }
+    }
+
+    private final MailRepositoryStore mailRepositoryStore;
+    private final Configuration configuration;
+    private final MailConverter mailConverter;
+
+    public MailRepositoryDeletedMessageVault(MailRepositoryStore mailRepositoryStore, Configuration configuration, MailConverter mailConverter) {
+        this.mailRepositoryStore = mailRepositoryStore;
+        this.configuration = configuration;
+        this.mailConverter = mailConverter;
+    }
+
+    @Override
+    public Publisher<Void> append(User user, DeletedMessage deletedMessage, InputStream inputStream) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(deletedMessage);
+        Preconditions.checkNotNull(inputStream);
+
+        MailRepository mailRepository = repositoryForUser(user);
+
+        return Mono.fromCallable(() -> mailRepository.store(mailConverter.toMail(deletedMessage, inputStream)))
```

Comment:
I didn't duplicate my comment about `subscribeOn` but it applies to all methods here I guess.